autometa/config/databases.py

```diff
@@ -66,230 +97,504 @@ class Databases:
+            raise ValueError(f'"section" must be "ncbi" or "markers". Provided: {section}')
+        if section == 'ncbi':
+            host = self.config.get(section, 'host')
+            ftp_fullpath = self.config.get('checksums', option)
+            chksum_fpath = ftp_fullpath.split(host)[-1]
+            with FTP(host) as ftp, tempfile.TemporaryFile() as fp:
+                ftp.login()
+                result = ftp.retrbinary(f'RETR {chksum_fpath}', fp.write)
+                if not result.startswith('226 Transfer complete'):
+                    raise ConnectionError(f'{chksum_fpath} download failed')
+                ftp.quit()
+                fp.seek(0)
+                checksum = fp.read().decode()
+        elif section == 'markers':
+            url = self.config.get('checksums', option)
+            with requests.Session() as session:
+                resp = session.get(url)
+                if not resp.ok:
+                    raise ConnectionError(f'Failed to retrieve {url}')
+                checksum = resp.text
+        return checksum
 
     def format_nr(self):
-        """Format NCBI nr.gz database into diamond formatted database nr.dmnd.
+        """Construct a diamond formatted database (nr.dmnd) from `nr` option
+        in `ncbi` section in user config.
+
+        NOTE: The checksum 'nr.gz.md5' will only be generated if nr.dmnd
+        construction is successful. If the provided `nr` option in `ncbi` is 'nr.gz'
+        the database will be removed after successful database formatting.
 
         Returns
         -------
         NoneType
             config updated option:'nr' in section:'ncbi'.
 
         """
-        db_infpath = self.config.get('ncbi', 'nr')
-        db_outfpath = db_infpath.replace('.gz', '.dmnd')
-        if not self.dryrun and not os.path.exists(db_outfpath):
-            diamond.makedatabase(
-                fasta=nr, database=db_infpath, nproc=self.nproc)
+        db_infpath = self.config.get('ncbi','nr')
+        db_infpath_md5 = f'{db_infpath}.md5'
+        db_outfpath = db_infpath.replace('.gz','.dmnd')
+        checksums_match = False
+        if os.path.exists(db_infpath_md5) and os.path.exists(db_outfpath):
+            # If nr.dmnd.md5 exists, then we will check if nr.gz.md5 is matching
+            current_checksum = read_checksum(db_infpath_md5)
+            current_hash, __ = current_checksum.split()
+            host = self.config.get('ncbi', 'host')
+            remote_checksum = self.get_remote_checksum(host, 'nr')
+            remote_hash, __ = remote_checksum.split()
+            if current_hash == remote_hash:
+                logger.debug(f'nr checksums match, skipping...')
+                checksums_match = True
+
+        no_update_db = os.path.exists(db_outfpath) and not self.do_update
+        if self.dryrun or no_update_db or checksums_match:
+            self.config.set('ncbi', 'nr', db_outfpath)
+            logger.debug(f'set ncbi nr: {db_outfpath}')
+            return
```

Comment:
After this point in the code we are here because `checksums_match == False`. So shouldn't we redownload the database? Or at least have some sort of option to do so. It seems weird that we check whether the checksums match and then don't do anything about it. Also, couldn't `checksums_match == True` but we still need to make the `nr.dmnd` file? It seems that in that scenario, the function will `return` and not format the database.

Reply:
1. `self.dryrun` is a user setting telling autometa _not_ to run the formatting/database downloading.
2. `no_update_db` (I've renamed to `do_not_upgrade`... Struggling to determine an appropriate variable name) checks that `nr.dmnd` exists _and_ the user does _not_ want to upgrade the database _even_ if the checksums do not match.
3. To toggle `checksums_match` to `True`, `os.path.exists(db_outfpath)` must also be `True`. So this ensures if the checksums match `nr.dmnd` will also exist.